### PR TITLE
Add documentation for overload names

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23844 Add documentation for overload names**

-

Differential Revision: [D16660167](https://our.internmc.facebook.com/intern/diff/D16660167/)

### CODE REVIEW

File: aten/src/ATen/native/README.md [hunk 1]
```diff
@@ -170,6 +170,10 @@ Note that argument type modifiers such as defaults and optional are not currentl
 
 The declarations also support the following attributes:
 
+**Overloads.** You can register multiple functions with the same name and different
```
Comment (dzhulgakov) [inline aten/src/ATen/native/README.md]: it renders weirdly (there's a preview button on the top right) as it's in the middle of the list of "attributes"

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/native/README.md (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/README.md]: this is an improvement, but it still leaves some unanswered questions.  For example: do I need to provide an overload if my function has the same name?  Is there a standard naming rule for the overload?

Comment (smessmer) [reply aten/src/ATen/native/README.md]: we don't have a fixed naming scheme defined yet, apart from "try to find a good, semantically meaningful name". Is your other question whether overload names need to be present and unique if the same function name appears multiple times? The text says "you can register multiple functions [...] *if* you give them unique overload names". So the answer would be yes. Should I change that to "only if"?

Comment (gchanan) [reply aten/src/ATen/native/README.md]: I think I'd be looking for more guidance.  It's not clear to be what I _should_ do, even if it says what is possible to do.

Comment (smessmer) [reply aten/src/ATen/native/README.md]: I added some more text. Is this better?

Comment (smessmer) [reply aten/src/ATen/native/README.md]: @dzhulgakov not sure what you mean with the rendering issue. Just checked and at least now (after making the text longer), I don't find what you mean. Is it fixed or still there?

Comment (dzhulgakov) [reply aten/src/ATen/native/README.md]: It's not rendering per-se, just weird that "The declarations also support the following attributes:" starts a list of attributes (with ### header), but the first thing is not an item in the list, but just a bold text. Maybe just reorder this paragraph above the previous one?

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/native/README.md [hunk 2]
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
...
+function signatures if you give them unique overload names. An overload name
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: what happens if you try to register the same function name without overloads?

File: aten/src/ATen/native/README.md [hunk 3]
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
...
+is specified after the function name, separated by a dot.
+
+Overload names do not have to be globally unique, but must be unique in the set
+of all overloads for the same function. Please try to make overload names
+semantically meaningful. An overload name that just enumerates the argument
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: You should expound on what makes a name semantically meaningful.  Just saying something like "the current standard is to name the first new parameter name or type" would be helpful.

File: aten/src/ATen/native/README.md [hunk 4]
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
...
+types isn't helpful. Actually it is hurtful, because overload names cannot be
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: I don't get it.  The above hints that I need to give an overload name ("You can register multiple functions with the same name and different function signatures if you give them unique overload names") -- okay technically it doesn't say if and only if...

But then here it tells me if I give it a bad name it's hurtful?  But how could it be worse than something that doesn't work?

File: aten/src/ATen/native/README.md [hunk 5]
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
...
+changed for backwards compatibility reasons, even if for some reason function
+argument types change.
+
+If you add a new overload to an existing function, please leave the existing
+overload names as they are (for backwards compatibility), but give the new
+overload a new, unique name.
+
+If you add a new function with multiple overloads, give them unique overload
+names, at most one overload is allowed to have no overload name and be the
+base variant.
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: base variant is not defined.

File: aten/src/ATen/native/README.md [hunk 6]
```diff
@@ -170,6 +170,27 @@ Note that argument type modifiers such as defaults and optional are not currentl
...
+**Overloads.** You can register multiple functions with the same name and different
+function signatures if you give them unique overload names. An overload name
+is specified after the function name, separated by a dot.
+
+Overload names do not have to be globally unique, but must be unique in the set
+of all overloads for the same function. Please try to make overload names
+semantically meaningful. An overload name that just enumerates the argument
+types isn't helpful. Actually it is hurtful, because overload names cannot be
+changed for backwards compatibility reasons, even if for some reason function
+argument types change.
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: wait, I can change the argument types but I can't change the overload name?  If that's the case this needs to be explained, i.e. what is allowed to change while still being the same overload?  (compatible types from python and new default parameters?)

**Review by gchanan — CHANGES_REQUESTED** [review]

**Pull request closed without merging**